amethyst_error/src/lib.rs

```diff
@@ -0,0 +1,402 @@
+    where
+        C: FnOnce(&Error) -> D,
+        D: Into<Error>;
+}
+
+impl<T, E> ResultExt<T> for result::Result<T, E>
+where
+    E: Into<Error>,
+{
+    fn with_context<C, D>(self, chain: C) -> Result<T, Error>
+    where
+        C: FnOnce(&Error) -> D,
+        D: Into<Error>,
+    {
+        match self {
+            Err(e) => {
+                let e = e.into();
+                let mut new = chain(&e).into();
+                new.source = Some(Box::new(e));
+                Err(new)
+            }
+            Ok(value) => Ok(value),
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub struct Causes<'a> {
+    current: Option<&'a Error>,
+}
+
+impl<'a> Iterator for Causes<'a> {
+    type Item = &'a Error;
+
+    fn next(&mut self) -> Option<Self::Item> {
+        if let Some(e) = self.current {
+            self.current = e.source();
+            return Some(e);
+        }
+
+        None
+    }
+}
+
+/// Constructs an `Error` using the standard string interpolation syntax.
+///
+/// ```rust
+/// #[macro_use] extern crate amethyst_error;
+///
+/// fn main() {
+///     let err = format_err!("number: {}", 42);
+///     assert_eq!("number: 42", err.to_string());
+/// }
+/// ```
+#[macro_export]
+macro_rules! format_err {
+    ($($arg:tt)*) => { $crate::Error::from_string(format!($($arg)*)) }
+}
+
+/// Constructs an [`Error`](Error) from a string.
+pub fn err_msg<M>(message: M) -> Error
+where
+    M: 'static + Send + Sync + fmt::Debug + fmt::Display,
+{
+    Error::new(ErrMsgError { message })
+}
+
+/// Treat something that can be displayed as an error.
+struct ErrMsgError<M> {
+    message: M,
+}
+
+impl<M> error::Error for ErrMsgError<M> where M: 'static + Send + Sync + fmt::Debug + fmt::Display {}
+
+impl<M> fmt::Display for ErrMsgError<M>
+where
+    M: fmt::Display,
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        self.message.fmt(fmt)
+    }
+}
+
+impl<M> fmt::Debug for ErrMsgError<M>
+where
+    M: fmt::Debug,
+{
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        self.message.fmt(fmt)
+    }
+}
+
+/// Test if backtracing is enabled.
+fn is_backtrace_enabled<F: Fn(&str) -> Option<ffi::OsString>>(get_var: F) -> bool {
+    match get_var(RUST_BACKTRACE) {
+        Some(ref val) if val != "0" => true,
+        _ => false,
+    }
+}
+
+fn new_backtrace() -> Option<Backtrace> {
+    if !HAS_BACKTRACE {
+        return None;
+    }
+
+    static ENABLED: atomic::AtomicUsize = atomic::ATOMIC_USIZE_INIT;
+
+    match ENABLED.load(atomic::Ordering::SeqCst) {
+        0 => {
+            let enabled = is_backtrace_enabled(|var| env::var_os(var));
+
+            ENABLED.store(enabled as usize + 1, atomic::Ordering::SeqCst);
+
+            if !enabled {
+                return None;
+            }
+        }
+        1 => return None,
+        _ => {}
+    }
+
+    Some(Backtrace::new())
+}
+
+const RUST_BACKTRACE: &str = "RUST_BACKTRACE";
+
+#[cfg(all(feature = "backtrace", feature = "std"))]
```

Comment:
This is for `no_std` support which is enabled through an attribute. I'm not sure you can make flags depend on it but I'll give it a shot!

Reply:
Why would Amethyst need `no_std`?

Reply:
Probably wouldn't. This is a requirement of the backtrace lib, but I can relax it.

The difference is that anything depending on no_std and amethyst_error _will_ blow up during compilation instead of degrading to not have backtraces.

Reply:
specs isn't going to work without an allocator.

Reply:
backtrace is no longer optional, thanks for the feedback!